aptos-move/aptos-memory-usage-tracker/src/lib.rs

```diff
@@ -0,0 +1,506 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_gas::{
+    AbstractValueSize, AptosGasMeter, AptosGasParameters, Fee, FeePerGasUnit, Gas,
+    GasScalingFactor, InternalGas, NumArgs, NumBytes,
+};
+use aptos_types::{
+    account_config::CORE_CODE_ADDRESS, contract_event::ContractEvent,
+    state_store::state_key::StateKey, write_set::WriteOp,
+};
+use move_binary_format::{
+    errors::{PartialVMError, PartialVMResult, VMResult},
+    file_format::CodeOffset,
+};
+use move_core_types::{language_storage::ModuleId, vm_status::StatusCode};
+use move_vm_types::{
+    gas::{GasMeter as MoveGasMeter, SimpleInstruction},
+    views::{TypeView, ValueView},
+};
+
+/// Special gas meter implementation that tracks the VM's memory usage based on the operations
+/// executed.
+///
+/// Must be composed with a base gas meter.
+pub struct MemoryTrackedGasMeter<G> {
+    base: G,
+
+    memory_quota: AbstractValueSize,
+    should_leak_memory_for_native: bool,
+}
+
+impl<G> MemoryTrackedGasMeter<G>
+where
+    G: AptosGasMeter,
+{
+    pub fn new(base: G) -> Self {
+        let memory_quota = base.gas_params().txn.memory_quota;
+
+        Self {
+            base,
+            memory_quota,
+            should_leak_memory_for_native: false,
+        }
+    }
+
+    #[inline]
+    fn use_heap_memory(&mut self, amount: AbstractValueSize) -> PartialVMResult<()> {
+        if self.feature_version() >= 3 {
+            match self.memory_quota.checked_sub(amount) {
+                Some(remaining_quota) => {
+                    self.memory_quota = remaining_quota;
+                    Ok(())
+                },
+                None => {
+                    self.memory_quota = 0.into();
+                    Err(PartialVMError::new(StatusCode::MEMORY_LIMIT_EXCEEDED))
+                },
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    #[inline]
+    fn release_heap_memory(&mut self, amount: AbstractValueSize) {
+        if self.feature_version() >= 3 {
+            self.memory_quota += amount;
+        }
+    }
+}
+
+// TODO: consider switching to a library like https://docs.rs/delegate/latest/delegate/.
+macro_rules! delegate {
+    ($(
+        fn $fn: ident $(<$($lt: lifetime),*>)? (&self $(, $arg: ident : $ty: ty)* $(,)?) -> $ret_ty: ty;
+    )*) => {
+        $(fn $fn $(<$($lt)*>)? (&self, $($arg: $ty),*) -> $ret_ty {
+            self.base.$fn($($arg),*)
+        })*
+    };
+}
+
+macro_rules! delegate_mut {
+    ($(
+        fn $fn: ident $(<$($lt: lifetime),*>)? (&mut self $(, $arg: ident : $ty: ty)* $(,)?) -> $ret_ty: ty;
+    )*) => {
+        $(fn $fn $(<$($lt)*>)? (&mut self, $($arg: $ty),*) -> $ret_ty {
+            self.base.$fn($($arg),*)
+        })*
+    };
+}
```

Comment:
Just a quick question: what'd these macros do? nit: for maintainability, it would probably be good to have some comments explaining the macro

Reply:
Very good point. I have some additional reorganization/refactoring coming down the road, and it would be more convenient if I add the comments as part of that.